### DISCUSSION TITLE
Driver for WaveShare 4.2 ePaper V2  with micropython on ESP32

### DISCUSSION THREAD

**Discussion by Barbsbou** [peterhinch/micropython-nano-gui#65] [Q&A]
Dear all;

I have a LOLIN D32 board and a new WS 4.2 white and Black ePaper on the 2nd version.

I am trying to use the Peter Hinch's driver to do so, I am using the pico_epaper_42.py file modifying the connection Pins and during the initialization step, `ssd=SSD()`  I am looping into the function `ready(self)` because the `(self._busy_pin() == 0)` is always true.

Any idea?????
Thanks a lot for your help!!!

**Comment by asherevan** [comment]
Did you try changing the pin used for the busy pin?
I’m not familiar with the particular hardware you are using, but it is something to try!

**Reply by Barbsbou** [reply]
Thank you for your reply!

Yes I tried different pin configurations follows the pinout scheme of my board, and I have still the problem...

Another idea?
Thank you in advance!

**Reply by Barbsbou** [reply]
Do you suspect an hardware problem with my pin?

**Comment by asherevan** [comment]
Possibly. As I said before, I’m not familiar with your particular hardware,
but it is possible.
Have you gotten the screen to work by other means? (Other board, driver,
etc.) And does the board work other than that problem?

**Reply by Barbsbou** [reply]
Indeed, I already used this board for different Proof Of Concepts, however it was always with less Pins...
So I will try with another!! I need a little bit time to process all necessary checks!

thank for your replies

**Comment by asherevan** [comment]
You're welcome!

**Comment by Barbsbou** [comment]
I tried with a new board, at the beginning, I had the same issue, so I changed the busy pin, and I don't have this problem.

All seems to be good from a software point of view, but the screen state does not change... nothing is printed on the screen....

**Comment by Barbsbou** [comment]
I read on the pinout on my board, and I used Pins for SPI. After readding hours, these Pins are used to connect/modify the firmware (or something like that)... Then I use Pins for V-SPI, and same problem, nothing happen to the epaper display...

I try to replace SPI with SoftSPI, but I am not sure to understand the difference between them...

**Comment by peterhinch** [comment]
I'm unclear what display is in use here. [This one](https://www.waveshare.com/pico-epaper-4.2.htm) is designed for the Pico but is supplied with a cable enabling other hosts to be connected. It has been extensively tested with particularly good results. However there is no guarantee of compatibility with other Waveshare displays of the same size.

I can only buy - and write drivers for - a limited number of these things and Waveshare make a vast range of products.

**Reply by Barbsbou** [reply]
First of all, THANK you a lot for all your shared sources, and I know that as soon as I will success to use my waveshare display I will use your other libraries!

OMG, indeed you are totally right: this is not the same hardware! I have [https://www.waveshare.com/wiki/4.2inch_e-Paper_Module_Manual](url) and if I understand correctly, the SPI communication schemes are different...
So I will search into your drivers list, which one is close to my hardware, any idea? I think the epd29driver.py?
Thus I will try to adapt it. If I will success I will share it.

See you from Normandy the 6th June ;-)

**Comment by Barbsbou** [comment]
Ok After hours of research and try to adapt the code for my display, and understand why it was not work:  the drivers proposed are for the V1 hardware, but now the hardware have different versions which change the communication protocol...

Moreover, I don't know if you know but Waveshare adapt the Peter Hinch's python-nano-gui to different hadware version. [Link corrected](https://github.com/waveshareteam/Pico_ePaper_Code/tree/main/pythonNanoGui)

For my case the adaptation is only for 1-bit monochrome. I will try to modify to adapt to 2-bits grayscale.
Thank you

**Reply by peterhinch** [reply]
That link is a 404.

**Reply by Barbsbou** [reply]
Oupsss  sorry

[Good link](https://github.com/waveshareteam/Pico_ePaper_Code/tree/main/pythonNanoGui)

I hope that I don't say anything new for you...

**Reply by peterhinch** [reply]
Thanks for the link. I wish Waveshare wouldn't do this: their copies are often out of date. It would be better if they linked to my repos.

**Reply by Barbsbou** [reply]
Sure it is a better practice...
